app/domain/authentication/authn_k8s/application_identity.rb

```diff
@@ -0,0 +1,171 @@
+        }
+      end
+
+      def container_name
+        annotation_name = "authentication-container-name"
+        annotation_value("authn-k8s/#{@service_id}/#{annotation_name}") ||
+          annotation_value("authn-k8s/#{annotation_name}") ||
+          annotation_value("kubernetes/#{annotation_name}") ||
+          "authenticator"
+      end
+
+      def namespace_scoped?
+        constraints.compact.empty?
+      end
+
+      private
+
+      # Validates that the application identity is defined correctly
+      def validate
+        validate_permitted_scope
+        validate_namespace_constraint_exists
+        validate_constraint_combinations
+      end
+
+      # If the application identity is defined in:
+      #   - annotations: validates that all the constraints are
+      #                  valid (e.g there is no "authn-k8s/blah" annotation)
+      #   - host id: validates that the host-id has 3 parts and that the given
+      #              constraint is valid (e.g the host id is not
+      #              "namesapce/blah/some-value")
+      def validate_permitted_scope
+        application_identity_in_annotations? ?
+          validate_permitted_annotations : validate_host_id
+      end
+
+      # This method assumes that 'validate_permitted_scope' has already run.
+      # It uses the value of 'host_id_suffix' after it is validated.
+      def validate_namespace_constraint_exists
+        return if namespace
+        raise Err::MissingNamespaceConstraint
+      end
+
+      # Validates that the application identity doesn't include logical constraint
+      # combinations (e.g deployment & deploymentConfig)
+      def validate_constraint_combinations
+        return if permitted_constraint_combinations?
+        raise Err::IllegalConstraintCombinations
+      end
+
+      def permitted_constraint_combinations?
+        # validate that we have only one constraint from the following
+        controllers = [:deployment, :deployment_config, :stateful_set]
+
+        (constraints.compact.keys & controllers).length <= 1
+      end
+
+      def constraint_value constraint_name
+        application_identity_in_annotations? ?
+          constraint_from_annotation(constraint_name) : constraint_from_id(constraint_name)
+      end
+
+      def constraint_from_annotation constraint_name
+        annotation_value("authn-k8s/#{@service_id}/#{constraint_name}") ||
+          annotation_value("authn-k8s/#{constraint_name}")
+      end
+
+      def annotation_value name
+        annotation = @host_annotations.find { |a| a.values[:name] == name }
+        annotation ? annotation[:value] : nil
+      end
+
+      def constraint_from_id constraint_name
+        host_id_suffix[-2] == constraint_name ? host_id_suffix[-1] : nil
+      end
+
+      def host_id_suffix
+        @host_name_suffix ||= @host_id.split('/').last(3)
+      end
+
+      def validate_permitted_annotations
+        validate_prefixed_permitted_annotations("authn-k8s")
```

Comment:
please test an host with following annotation:
authn-k8s/namspace = blah
authn-k8sInvalid/namspace = blah
what is the expected behavior ?

Reply:
This is the same test as the above. This annotation will be ignored.

Reply:
Added a UT: https://github.com/cyberark/conjur/blob/host-annotations/spec/app/domain/authentication/authn_k8s/application_identity_spec.rb#L348